### ISSUE TITLE
return=representation returns singular object if you insert an array with 1 element

### ISSUE THREAD

**Issue by yuhe00** [begriffs/postgrest#753]
The following works as expected:
```
echo '[{"name":"Hyundai"},{"name":"Mazda"}]' | http POST localhost:3337/makes "Prefer:return=representation"
HTTP/1.1 201 Created
Content-Encoding: gzip
Content-Type: application/json; charset=utf-8
Date: Mon, 28 Nov 2016 09:58:49 GMT
Server: postgrest/0.3.2.0
Transfer-Encoding: chunked

[
    {
        "id": "f4a99b15-2144-4563-8b5a-41a1456fb692", 
        "name": "Hyundai", 
        "status": "UNPUBLISHED"
    }, 
    {
        "id": "21205052-77a8-4e1c-b78d-a49c42c41307", 
        "name": "Mazda", 
        "status": "UNPUBLISHED"
    }
]
```
...while the following returns a single object:
```
> echo '[{"name":"Hyundai"}]' | http POST localhost:3337/makes "Prefer:return=representation"
HTTP/1.1 201 Created
Content-Encoding: gzip
Content-Type: application/json; charset=utf-8
Date: Mon, 28 Nov 2016 09:52:20 GMT
Location: /makes?id=eq.f6adee2d-c2a2-4bf9-afbb-d8ecc6ead2a6
Server: postgrest/0.3.2.0
Transfer-Encoding: chunked

{
    "id": "f6adee2d-c2a2-4bf9-afbb-d8ecc6ead2a6", 
    "name": "Hyundai", 
    "status": "UNPUBLISHED"
}
```
Not sure if this is intended behavior, but it was a bit unexpected. I think it should return an array if you send an array, no matter the amount of elements in the array.

**Comment by ruslantalpa** [comment]
This is happening [here](https://github.com/begriffs/postgrest/blob/master/src/PostgREST/App.hs#L136), not sure why that was decided

**Comment by ruslantalpa** [comment]
i think this is just a remanence from the fact that you can post like this
`{"name":"Hyundai"}`  and also like this `[{"name":"Hyundai"},{"name":"Mazda"}]`
we'd need to look in the old issues to see if it was an oversight or there is a specific reason

**Comment by NikolayS** [comment]
Hi, we have the same issue.

In my case, frontend developer was very confused after he saw JSON object in the reponse's body, not an array. He passed a single JSON object to the POST request. I suspected that to have an array the response, he just needs to add `[ .. ]` (use JSON array in request). But it didn’t help — if we pass JSON array but with single method, response has single JSON object, not array. So it’s hard to create universal processing code on frontend part.

At the same time, DELETE always returns JSON array, so it overall looks like inconsistency.

I would expect:
  - PATCH: if the request has JSON object, response has also JSON object. Otherwise, both have JSON array;
 - DELETE always returns array, even if only single row was deleted, because it's hard to predict how many rows will be really deleted.

This is what I'd expect from my user point of view, but maybe I'm missing something essential.

**Comment by NikolayS** [comment]
Also, I noticed two more "maybe-inconsistent" things:

 - PATCH of single row returns JSON array, although I requested to update single row (using PK). Again, maybe it's OK, because it's predict -- frontend code might have no idea, what are PK/UKs at all;
 - PATCH returns empty array if no rows affected, at the same time, DELETE returns empty **body** (not empty JSON array) if no rows affected - again, this looks like an inconsistency.

**Comment by NikolayS** [comment]
_(notice: I edited my very first comment here, it was talking about PATCHes, not POSTs by mistake. Now it's correct)_

More about inserting (POSTs). If we use single JSON in the request, we would expect single row created (in case of success) and having single JSON object in the response.

But look at this example (yes it's ugly, but anyway): https://gist.github.com/NikolayS/61be5662e76277385234e3857436d5a8

If we try to insert single row, two rows will be created and returned via `returning *`.

PostgREST fails in this case:
```http
POST /multi HTTP/1.1
Host: localhost
Content-Type: application/json
Prefer: return=representation
Cache-Control: no-cache

{"data": "manual"}
```
gives us `201 Created` http status/code, and "Syntax error" (in plain text!! with `Content-Type: application/json; charset=utf-8` header) in the response's body.

Bottom line: maybe it's better to expect multiple rows always being INSERTed, even if user asked to insert into one -- therefore maybe it's better to reply with JSON array always. 
It is really hard to predict, how many rows will be affected in case of all operations -- insert/update/delete. Mutating response format based on real cardinality taken from Postgres' actions is not a good idea, since it will make working with API more difficult. So I vote for "JSON array always".

**Comment by begriffs** [comment]
Sorry for the inconsistency. Here's my proposal:

* If the request specifies `Accept: application/vnd.pgrst.object` (or whatever mime type we end up choosing in #763)
  * If the response would have been a singleton array
    * respond with single JSON object 
  * Else
    * respond with HTTP 406 (Not Acceptable) and roll back the transaction for updated rows
* Else if `application/json` or no Accept header specified
  * all verbs respond with json arrays (even inserts with json object as payload)

Let me know if this is how you'd like it to behave and I'll incorporate this change into pull request 763.

**Comment by NikolayS** [comment]
Looks good! 
And what about cases when no rows affected? Currently, as I mentioned above, we have in such cases:
 - PATCH returns empty JSON array if no rows affected, at the same time, 
 - DELETE returns empty http body (not an empty JSON array)

For POST, it's also theoretically possible that no rows will be inserted – consider a "before" trigger that doesn't pass NEW record further:
```sql
create or replace function nopasaran() returns trigger as $$
begin return null; end;
$$ language plpgsql;
create trigger t_nopasaran before insert on multi for each row execute procedure nopasaran();
```

It work without any errors in Postgres: It will accept any INSERT doing nothing, and returning empty result in Postgres. But for PostgREST, it will be `500 Internal Server Error` with this body:
```json
{
  "details": "Row number 0",
  "message": "Row error: unexpected null"
}
```

**Comment by begriffs** [comment]
Good question about the zero-row responses. Here's what I'm imagining:

* Requesting `application/vnd.pgrst.object` returns http 406 when response has zero rows.
  * Also have `vnd.pgrst.object` fail for any number of updated/deleted rows (even exactly one row) when request lacks an accompanying `Prefer: return=representation` because a blank result is not an object.
* Requesting `application/json` returns `[]` for reads with zero rows or updates *with* `Prefer: return=representation` which update no rows.

How does this sound? Can you see any weird interactions with these rules, or am I missing any cases?

**Comment by NikolayS** [comment]
Currently, if you try to delete or update a non-existing row, you receive 404 Not Found. This is straightforward and reflects the fact that what are you trying to find doesn't exist.
406 code will conflict with this logic, right? From API user's point of view, 404 looks like more expected code when smth is not found.

BTW, it doesn't directly relates to this discussion, but it seems that I found another problem -- when I try to DELETE some row in a table, for which I don't have "delete" permissions, I get `"permission denied for relation tablename"`, which is fine, but the http code is 404... I'd expect something different, maybe 403.

**Comment by NikolayS** [comment]
The same for PATCH – when we have `permission denied for relation...` response, it's also with 404 code.

**Comment by NikolayS** [comment]
Oh, 404 issue is described in #575 already.

**Comment by begriffs** [comment]
I believe the inconsistent responses are fixed in the 0.4 release. Please re-open if you find a problem.
